From 18dab341a8986d53a8bd417ff77c733c70266a2c Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:46:38 -0400
Subject: [iommu] vt-d: Free resources if failed to create domain for PCIe endpoint

Message-id: <20140905014638.31297.9678.stgit@gir.stowe>
Patchwork-id: 91300
O-Subject: [RHEL7.1 PATCH 08/65] iommu/vt-d: Free resources if failed to create domain for PCIe endpoint
Bugzilla: 1129880 1087643
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1087643
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129880
Upstream Status: e85bb5d4d1d5ef55cf4fbb4da5dda67d2bbacc5c

commit e85bb5d4d1d5ef55cf4fbb4da5dda67d2bbacc5c
Author: Jiang Liu <jiang.liu@linux.intel.com>
Date:   Wed Feb 19 14:07:27 2014 +0800

    iommu/vt-d: Free resources if failed to create domain for PCIe endpoint

    Enhance function get_domain_for_dev() to release allocated resources
    if failed to create domain for PCIe endpoint, otherwise the allocated
    resources will get lost.

    Signed-off-by: Jiang Liu <jiang.liu@linux.intel.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/intel-iommu.c b/drivers/iommu/intel-iommu.c
index 67b114e..503cc73 100644
--- a/drivers/iommu/intel-iommu.c
+++ b/drivers/iommu/intel-iommu.c
@@ -2015,7 +2015,7 @@ static int dmar_insert_dev_info(int segment, int bus, int devfn,
 /* domain is initialized */
 static struct dmar_domain *get_domain_for_dev(struct pci_dev *pdev, int gaw)
 {
- struct dmar_domain *domain;
+ struct dmar_domain *domain, *free = NULL;
  struct intel_iommu *iommu;
  struct dmar_drhd_unit *drhd;
  struct pci_dev *dev_tmp;
@@ -2062,17 +2062,16 @@ static struct dmar_domain *get_domain_for_dev(struct pci_dev *pdev, int gaw)
   free_domain_mem(domain);
   goto error;
  }
- if (domain_init(domain, gaw)) {
-  domain_exit(domain);
+ free = domain;
+ if (domain_init(domain, gaw))
   goto error;
- }
 
  /* register pcie-to-pci device */
  if (dev_tmp) {
-  if (dmar_insert_dev_info(segment, bus, devfn, NULL, &domain)) {
-   domain_exit(domain);
+  if (dmar_insert_dev_info(segment, bus, devfn, NULL, &domain))
    goto error;
-  }
+  else
+   free = NULL;
  }
 
 found_domain:
@@ -2080,6 +2079,8 @@ found_domain:
      pdev, &domain) == 0)
   return domain;
 error:
+ if (free)
+  domain_exit(free);
  /* recheck it here, maybe others set it */
  return find_domain(pdev);
 }
-- 
1.7.1